lib/Dialect/HW/Transforms/LowerHWtoBTOR2.cpp

```diff
@@ -0,0 +1,786 @@
+    genSort(bitvec, 1);
+
+    // Check for an overaching enable
+    // In our case the sv.if operation will probably only be used when
+    // conditioning an sv.assert on an enable signal. This means that
+    // its condition is probably used to imply our assertion
+    if (auto ifop = dyn_cast<sv::IfOp>(((Operation *)op)->getParentOp())) {
+      Value en = ifop.getOperand();
+
+      // Generate the implication
+      genImplies(ifop, en, expr);
+
+      // Generate the implies inversion
+      genUnaryOp(op, ifop, notStr, 1);
+    } else {
+      // Generate the expression inversion
+      genUnaryOp(op, expr, notStr, 1);
+    }
+
+    // Genrate the bad btor2 intruction
+    genBad(op);
+  }
+  // Assumptions are converted to a btor2 constraint instruction
+  void visit(sv::AssumeOp op) {
+    // Extract the expression that we want our constraint to be about
+    Value expr = op.getExpression();
+    genConstraint(expr);
+  }
+
+  // Firrtl registers generate a state instruction
+  // The final update is also used to generate a set of next btor
+  // instructions
+  void visit(seq::FirRegOp reg) {
+    // Start by retrieving the register's name and width
+    StringRef regName = reg.getName();
+    int64_t w = requireSort(reg.getType());
+
+    // Generate state instruction (represents the register declaration)
+    genState(reg, w, regName);
+
+    // Record the operation for future `next` instruction generation
+    // This is required to model transitions between states (i.e. how a
+    // register's value evolves over time)
+    regOps.push_back(reg);
+  }
+};
+} // end anonymous namespace
+
+void LowerHWtoBTOR2Pass::runOnOperation() {
+  // Start by checking for each module in the circt, we only consider
+  // the 1st one, as btor2 does not have the concept of modules or module
+  // hierarchies. We assume that no nested modules exist at this point.
+  // This greatly simplifies translation.
+  getOperation().walk([&](hw::HWModuleOp module) {
+    // Start by extracting the inputs and generating appropriate instructions
+    for (auto &port : module.getPortList()) {
+      visit(port);
+    }
+
+    // Visit all of the operations in our module
+    module.walk([&](Operation *op) {
+      TypeSwitch<Operation *, void>(op)
+          .template Case<
+              // All supported hw operations
+              hw::OutputOp, hw::ConstantOp, hw::WireOp,
+              // All supported comb ops
+              comb::AddOp, comb::SubOp, comb::MulOp, comb::DivUOp, comb::DivSOp,
+              comb::ModSOp, comb::ShlOp, comb::ShrUOp, comb::ShrSOp,
+              comb::AndOp, comb::OrOp, comb::XorOp, comb::ConcatOp,
+              comb::ExtractOp, comb::ICmpOp, comb::MuxOp,
+              // All supported sv operations
+              sv::AssertOp, sv::AssumeOp,
+              // All supported seq operations
+              seq::FirRegOp>([&](auto expr) { visit(expr); })
+          // Ignore anything else
+          .Default([&](auto expr) {});
```

Comment:
Can we emit an error for unhandled operations (e.g. instance, array/struct operations)?